### PR TITLE
EXZ.2 initial submission

### PR DESCRIPTION
**Pull request by jni2000**
Initial submission of EXZ.2. Please review

### CODE REVIEW

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 1]
```diff
@@ -294,3 +294,263 @@ def test_WINNF_FT_S_EXZ_1(self,config_filename):
...
+    grants_N1 = [grant_N1_1, grant_N1_2, grant_N1_3]
+    grants_N2 = [grant_N2_1, grant_N2_2]
+    grants_N3 = [grant_N3_1, grant_N3_2, grant_N3_3]
+
+    devices_N1 = [device_N1_1, device_N1_2, device_N1_3]
+    devices_N2 = [device_N2_1, device_N2_2]
+    devices_N3 = [device_N3_1, device_N3_2, device_N3_3]
+
+    conditionals = {
+        'registrationData': [conditionalsN1, conditionalsN2, conditionalsN3]
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Please update to be consistent with the latest code in master.

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 2]
```diff
@@ -294,3 +294,263 @@ def test_WINNF_FT_S_EXZ_1(self,config_filename):
...
+    # Very light checking of the config file.
+    self.assertEqual(len(config['registrationRequestsN1']),len(config['grantRequestsN1']))
+    self.assertEqual(len(config['registrationRequestsN2']),len(config['grantRequestsN2']))
+    self.assertEqual(len(config['registrationRequestsN3']),len(config['grantRequestsN3']))
+
+    # Enforce NTIA Exlusion zones
+    self._sas_admin.TriggerEnableNtiaExclusionZones()
+
+    # Pre-load conditional registration data for N1,N2 and N3 CBSDs.
+    if ('conditionalRegistrationData' in config) and (config['conditionalRegistrationData']):
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Please update this as well.

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 3]
```diff
@@ -294,3 +294,263 @@ def test_WINNF_FT_S_EXZ_1(self,config_filename):
...
+    # Register N3 devices
+    cbsd_ids_N3 = self.assertRegistered(config['registrationRequestsN3'])
+
+
+    # Trigger daily activities
+    self.TriggerDailyActivitiesImmediatelyAndWaitUntilComplete()
+
+    # Generating grant requests for N1 and validating responses
+    grant_request_N1 = []
+    for index,grants in enumerate(config['grantRequestsN1']):
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: You can use `util.addCbsdIdsToRequests()` instead.

**Comment by jni2000** [comment]
Kate

We have addressed your latest review comments. 

For the following comments

> kate-harrison 10 hours ago 
Contributor
Please update to be consistent with the latest code in master.

We found that the latest master branch code also in the same form as we have coded, however we did see a “ PR479 - Fix conditionalRegistrationData” ,
In which conditional parameter changes are proposed, so we made the changes as per PR479 by assuming reviewer also referring the same.  

Please continue review.

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 4]
```diff
@@ -456,63 +452,55 @@ def generate_EXZ_2_default_config(self, filename):
...
     # Enforce NTIA Exlusion zones
     self._sas_admin.TriggerEnableNtiaExclusionZones()
 
-    # Pre-load conditional registration data for N1,N2 and N3 CBSDs.
-    if ('conditionalRegistrationData' in config) and (config['conditionalRegistrationData']):
-      self._sas_admin.PreloadRegistrationData(config['conditionalRegistrationData'])
-
     # Register N1 devices
-    cbsd_ids_N1 = self.assertRegistered(config['registrationRequestsN1'])
+    cbsd_ids_N1 = self.assertRegistered(config['registrationRequestsN1'], config['conditionalRegistrationData'][0])
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: As suggested defined separate conditionals for N1, N2 and N3 CBSD’s

**Comment by jni2000** [comment]
> Kate:
It's not totally clear to me exactly which exclusion zones are referred to in this test case. I want to get a little more clarity before submitting, but otherwise this test case is very nearly ready to submit. I just sent an email to WG4 to ask for their help on this matter.

We use the NTIA Exclusion zone information available under the link “https://www.ntia.doc.gov/fcc-filing/2015/ntia-letter-fcc-commercial-operations-3550-3650-mhz-band”.   The CBSD locations for the sets N1, N2 and N3 were chosen based on the ‘Ground Based Exclusion Zones‘, ‘Shipborne Radar Envelope Exclusion Zones’  and ‘Dynamic Protection Areas’ details provided under that link.
The below are details of CBSD locations w.r.t. to NTIA zones:

N1 CBSDs each located outside 50 meters of below Exclusion Zones
device_N1_1 - Fort Carson Military Reservation
device_N1_2 - White Sands Missile Range
device_N1_3 - Yuma Proving Ground

N2 CBSDs each located within below Exclusion Zones
device_N2_1 - Fort Carson Military Reservation
device_N2_2 - White Sands Missile Range

N3 CBSDs each located within 50 meters of below Exclusion Zones
device_N3_1 - White Sands Missile Range
device_N3_2 - Yuma Proving Ground
device_N3_3 - Fort Irwin

**Comment by kate-harrison** [comment]
I am checking via email with Masoud whether these are valid sites, but it seems they may not be since only some zones listed in the doc will actually be exclusion zones.

In the meantime, please chat with Kambiz to understand the changes that were made to EXZ.1 and EXZ.2 this morning. (The good news is that they only require you to remove code :) )

**Comment by jni2000** [comment]
Kate

Here is the update I got from Masoud with regard to EXZ.2
1.  EXZ.2 will be only required for GB1 and GB2 inland radars, not required for coastal exclusion zones which will be covered by DPAs.
2. The NTIA web site does not have the whole list of GB1 and GB2 inland radar exclusion zone boundary yet, but some of them are available.
3. we have the whole list of the GB1 and GB2 site information in spreadsheet.

Questions to you are:
1. do we need the whole inland radar exclusion zones for this EXZ.2 implementation ? My feeling is not.
2. are the answers from Masoud above good enough for us to move on and continue the EXZ.2 implementation? My feeling on this is yes.

Please comment. Thanks.

**Comment by kate-harrison** [comment]
James, I checked with some of our government friends and they seem to have a different interpretation than Masoud. I'm waiting for confirmation and then I will pass on the information, so please sit tight for a few more days. I think the implementation is simple enough that we can finalize very quickly (in either direction) so I don't expect a negative impact to the schedule.

**Comment by jni2000** [comment]
ok, thanks.

**Comment by sbdt-google** [comment]
If it can help, some time ago, I have put the latest GBS exclusion zone KML definition in the repository, with routine
```
 zones.py:   GetExclusionZones()
```
The file will be updated if they give us an update to it

**Comment by jni2000** [comment]
Thanks, Serge. We are waiting for clarifications from the DoD/NTIA/ITS as Kate mentioned.

**Comment by kate-harrison** [comment]
I confirmed with Rebecca this morning that EXZ.2 is meant to test the coastal exclusion zones (i.e. the exclusion zones which will become obsolete once we have a waiver from the FCC saying that they can be implemented as DPAs). So I think we now have enough information to implement, but please let me know if you have questions.

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 5]
```diff
@@ -213,10 +213,10 @@ def generate_EXZ_1_default_config(self, filename):
   def test_WINNF_FT_S_EXZ_1(self,config_filename):
     """Injected Exclusion Zones
 
-    The response should be SUCCESS for CBSDs located outside 50meters of all 
+    The response should be SUCCESS for CBSDs located outside 50meters of all
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Removed   the comment

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 6]
```diff
@@ -520,11 +532,6 @@ def test_WINNF_FT_S_EXZ_2(self, config_filename):
     response_N2 = self._sas.Grant(request_N2)['grantResponse']
     self.assertEqual(len(response_N2), len(grant_request_N2))
 
-    for response_num, response in enumerate(response_N2):
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Added   check for N2 and removed N1 checks

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 7]
```diff
@@ -294,53 +289,71 @@ def test_WINNF_FT_S_EXZ_1(self,config_filename):
...
-    # Load device info
-    # Move the device locations outside 50 meters of all NTIA exclusion zones
+
+    # Loading N1 set of CBSDs each located outside 50 meters of all Exclusion Zones
     device_N1_1 = json.load(
         open(os.path.join('testcases', 'testdata', 'device_a.json')))
-    device_N1_1['installationParam']['latitude'] = 35.63887
-    device_N1_1['installationParam']['longitude'] = -102.68289
+    # Moving device_N1_1 just outside of 50 meters from exclusion zone of
+    # 'Fort Carson Military Reservation'
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Changed   the location of the CBSDs to align with the coastal exclusion zone coverage

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 8]
```diff
@@ -452,16 +465,16 @@ def generate_EXZ_2_default_config(self, filename):
...
         'registrationRequestsN3': devices_N3,
         'grantRequestsN1': grants_N1,
         'grantRequestsN2': grants_N2,
         'grantRequestsN3': grants_N3,
-        'conditionalRegistrationData': conditionals
+        'conditionalRegistrationDataN1': [conditionals_N1],
+        'conditionalRegistrationDataN2': [conditionals_N2],
+        'conditionalRegistrationDataN3': [conditionals_N3]
     }
     writeConfig(filename, config)
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Removed   the comment

**Comment by jni2000** [comment]
Kate, thanks for all the comments. We have addressed them. Please take a look.

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 9]
```diff
@@ -293,4 +285,258 @@ def test_WINNF_FT_S_EXZ_1(self,config_filename):
...
+    grant_N1_3['operationParam']['operationFrequencyRange'][
+        'lowFrequency'] = 3645000000
+    grant_N1_3['operationParam']['operationFrequencyRange'][
+        'highFrequency'] = 3655000000
+
+    # Forming N2 grant request
+    grant_N2_1 = json.load(
+          open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_N2_1['operationParam']['operationFrequencyRange'][
+        'lowFrequency'] = 3650000000
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Do the coastal exclusion zones extend above 3650 MHz?

Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: fixed now.

**Comment by jni2000** [comment]
Kate, we have fixed the the frequency range issue as you pointed out, please take a look. Thanks.

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 10]
```diff
@@ -293,4 +285,258 @@ def test_WINNF_FT_S_EXZ_1(self,config_filename):
...
+    device_N3_2['installationParam']['latitude'] = 47.35184
+    device_N3_2['installationParam']['longitude'] = -123.68957
+    device_N3_2['fccId'] = "test_fcc_id_bbb"
+    device_N3_2['cbsdSerialNumber'] = "test_serial_number_bbb"
+
+    device_N3_3 = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_c.json')))
+    # Moving device_N3_3 to a location within 50 meters from the exclusion zone
+    # of 'East-Gulf Combined Contour'
+    device_N3_3['installationParam']['latitude'] = 37.78916
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: device_N3_3, device_N3_2, device_N3_1 are all within 50 "kilometers" and not 50 "meters". Could you pls double check these 3 device locations. Thanks.

Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Thanks, fixed.

**Comment by jni2000** [comment]
harinineelu

Thanks for reviewing, we have fixed the N3 location. Please take a look.

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 11]
```diff
@@ -293,4 +285,258 @@ def test_WINNF_FT_S_EXZ_1(self,config_filename):
...
+    device_N3_2['installationParam']['latitude'] = 47.35184
+    device_N3_2['installationParam']['longitude'] = -123.68957
+    device_N3_2['fccId'] = "test_fcc_id_bbb"
+    device_N3_2['cbsdSerialNumber'] = "test_serial_number_bbb"
+
+    device_N3_3 = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_c.json')))
+    # Moving device_N3_3 to a location within 50 meters from the exclusion zone
+    # of 'East-Gulf Combined Contour'
+    device_N3_3['installationParam']['latitude'] = 39.78916
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Hi James, this location is not within 50 meters from the exclusion zone too. 
Also, device_N3_2 and device_N3_1 need to be changed. Could you please recheck. Thanks.

Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: Sorry, they shall be fixed now. Please check.

**Comment by jni2000** [comment]
harinineelu

We fixed all three locations to be within 50 miters of EXZ. please check. Thanks.

**Comment by harinineelu** [comment]
device_N3_2 looks good. Could you pls consider using the suggested points for device_N3_1 and 3. Thanks!

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 12]
```diff
@@ -334,26 +334,26 @@ def generate_EXZ_2_default_config(self, filename):
         open(os.path.join('testcases', 'testdata', 'device_a.json')))
     # Moving device_N3_1 to a location within 50 meters from the exclusion zone
     # of 'East-Gulf Combined Contour'
-    device_N3_1['installationParam']['latitude'] = 43.94141
-    device_N3_1['installationParam']['longitude'] = -71.50314
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: changed as suggested

File: src/harness/testcases/WINNF_FT_S_EXZ_testcase.py [hunk 13]
```diff
@@ -334,26 +334,26 @@ def generate_EXZ_2_default_config(self, filename):
...
+    device_N3_2['installationParam']['longitude'] = -124.15241
     device_N3_2['fccId'] = "test_fcc_id_bbb"
     device_N3_2['cbsdSerialNumber'] = "test_serial_number_bbb"
 
     device_N3_3 = json.load(
         open(os.path.join('testcases', 'testdata', 'device_c.json')))
     # Moving device_N3_3 to a location within 50 meters from the exclusion zone
     # of 'East-Gulf Combined Contour'
-    device_N3_3['installationParam']['latitude'] = 39.78916
-    device_N3_3['installationParam']['longitude'] = -124.03314
```
[the remark this answers is not in the document]
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_EXZ_testcase.py]: changed as suggested.

**Comment by jni2000** [comment]
harinineelu

We have changed the location of devices N3_1 and N3_3 as you suggested. Please take a look. Thanks.

**Comment by harinineelu** [comment]
Sorry, I made a mistake with the device_N3_1 location. The right lat/long needs to be 43.936508, -71.098682. 

Both device_N3_2 and device_N3_3, look good now.

**Comment by jni2000** [comment]
harinineelu

Thanks for the checking, N3_1 is now updated to (43.936508, -71.098682).

**Comment by harinineelu** [comment]
Great, Looks good.

**Pull request merged**